File: ios-app/TomsDiary/Models/LLMConfig.swift
//
//  LLMConfig.swift
//  Tom's Diary - iOS
//
//  Configuration for the LLM (OpenAI-compatible API).
//  The app can use any OpenAI-compatible endpoint including:
//  - Local vLLM instances
//  - Ollama
//  - Any other OpenAI-compatible API
//

import Foundation

struct LLMConfig {
    
    // MARK: - Properties
    
    let baseUrl: String
    let apiKey: String
    let model: String
    
    // MARK: - Defaults
    
    static let `default` = LLMConfig(
        baseUrl: "http://localhost:8001/v1",
        apiKey: "placeholder",
        model: "default"
    )
    
    // MARK: - Initialization
    
    init(
        baseUrl: String = "http://localhost:8001/v1",
        apiKey: String = "placeholder",
        model: String = "default"
    ) {
        self.baseUrl = baseUrl
        self.apiKey = apiKey
        self.model = model
    }
    
    // MARK: - Persistence
    
    private enum Keys {
        static let baseUrl = "llm_base_url"
        static let apiKey = "llm_api_key"
        static let model = "llm_model"
    }
    
    static func load(from userDefaults: UserDefaults = .standard) -> LLMConfig {
        LLMConfig(
            baseUrl: userDefaults.string(forKey: Keys.baseUrl) ?? `default`.baseUrl,
            apiKey: userDefaults.string(forKey: Keys.apiKey) ?? `default`.apiKey,
            model: userDefaults.string(forKey: Keys.model) ?? `default`.model
        )
    }
    
    func save(to userDefaults: UserDefaults = .standard) {
        userDefaults.set(baseUrl, forKey: Keys.baseUrl)
        userDefaults.set(apiKey, forKey: Keys.apiKey)
        userDefaults.set(model, forKey: Keys.model)
    }
    
    // MARK: - Validation
    
    func validate() -> Bool {
        // Validate URL
        guard let url = URL(string: baseUrl),
              url.scheme == "http" || url.scheme == "https" else {
            return false
        }
        
        // URL and model should not be empty
        return !baseUrl.isEmpty && !model.isEmpty
    }
}

// MARK: - Codable

extension LLMConfig: Codable {
    enum CodingKeys: String, CodingKey {
        case baseUrl = "base_url"
        case apiKey = "api_key"
        case model
    }
}

File: android-app/app/src/main/java/com/tomsdiary/LLMConfig.kt
package com.tomsdiary

import android.content.Context
import android.content.SharedPreferences

/**
 * Configuration for the LLM (OpenAI-compatible API).
 * 
 * The app can now use any OpenAI-compatible endpoint, including:
 * - Local vLLM instances (http://localhost:8001/v1)
 * - Ollama (http://localhost:11434/v1)
 * - Any other OpenAI-compatible API
 * 
 * Update these values in the app settings or modify the defaults here.
 */
object LLMConfig {
    
    // Default configuration - can be overridden in SharedPreferences
    const val DEFAULT_BASE_URL = "http://localhost:8001/v1"
    const val DEFAULT_API_KEY = "placeholder"  // Most local LLMs don't require a real key
    const val DEFAULT_MODEL = "default"
    
    fun getBaseUrl(context: Context): String {
        val prefs = context.getSharedPreferences("toms_diary", Context.MODE_PRIVATE)
        return prefs.getString("llm_base_url", DEFAULT_BASE_URL) ?: DEFAULT_BASE_URL
    }
    
    fun getApiKey(context: Context): String {
        val prefs = context.getSharedPreferences("toms_diary", Context.MODE_PRIVATE)
        return prefs.getString("llm_api_key", DEFAULT_API_KEY) ?: DEFAULT_API_KEY
    }
    
    fun getModel(context: Context): String {
        val prefs = context.getSharedPreferences("toms_diary", Context.MODE_PRIVATE)
        return prefs.getString("llm_model", DEFAULT_MODEL) ?: DEFAULT_MODEL
    }
    
    fun setBaseUrl(context: Context, baseUrl: String) {
        val prefs = context.getSharedPreferences("toms_diary", Context.MODE_PRIVATE)
        prefs.edit().putString("llm_base_url", baseUrl).apply()
    }
    
    fun setApiKey(context: Context, apiKey: String) {
        val prefs = context.getSharedPreferences("toms_diary", Context.MODE_PRIVATE)
        prefs.edit().putString("llm_api_key", apiKey).apply()
    }
    
    fun setModel(context: Context, model: String) {
        val prefs = context.getSharedPreferences("toms_diary", Context.MODE_PRIVATE)
        prefs.edit().putString("llm_model", model).apply()
    }
}
